torch/_inductor/ir.py

```diff
@@ -3984,6 +3984,123 @@ def codegen(self, wrapper):
         wrapper.writeline(f"_register_tensor_work({output_name}, {output_name}_work)")
 
 
+class MultiOutputNoSizeAssert(MultiOutput):
+    """
+    Extract partial output from a multi-output OP.
+        Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
+    """
+
+    def codegen(self, wrapper):
+        wrapper.writeline(
+            f"{self.get_name()} = {self.inputs[0].get_name()}{self.index}"
+        )
+
+
+class ForceInPlace(ExternKernel):
+    """
+    Helper OP to encode an in/out argument that tries to make it inplace whenever possible.
+    Wrap the input of your inplace op to enable this behavior.
+
+    TODO: We should test whether wait_tensor can be a victim of reordering and lead to data races.
```

Comment:
you should at least test that in the simple cases this does not happen.  See my tests in `test_inductor_collectives.py` which use FileCheck to inspect the relative order of nodes in output codegen.

Reply:
This comment is actually stale as the final design is to emit wait_tensor calls for all of the resulting tensors